docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+operations to be performed by a service prove their ability to obtain a _device
+key_ signature (see [Signatures](#signatures)) themselves. This provides basic
+protection against unauthorized access, as well as accidental routing errors.
+
+To achieve this, we devise a simple protocol:
+
+RPC messages are wrapped in either a `request` or `response` envelope structure
+as defined below:
+
+```cddl
+request = (
+    request-headers,
+    ? payload: bstr,
+)
+
+response = ok / error
+
+ok = (
+    response-headers,
+    ? payload: bstr,
+)
+
+error = (
+    response-headers,
+    code: error-code,
+    ? message: tstr,
+)
+
+request-headers = (
+    client-id: tstr .size (4..16),
+    ? ticket: ticket,
+    ? signature: bstr .size 64,
+)
+
+response-headers = (
+    ? request: ticket,
+    ? next: ticket,
+)
+
+error-code = forbidden / connection-close
+
+ticket = bstr .size (16..64)
+```
+
+> Note: for brevity we use [CDDL] groups (parenthesis) to denote [array
+> encoding][cbor-array]. Embedded groups (without a label) imply that the group
+> is inlined.
+
+The `client-id` is an unambiguous, human-readable string identifying the client
+application (e.g. "radicle-cli").
+
+The `payload` is the CBOR-encoded application-level message.
+
+A `ticket` serves as a request-id (enabling pipelining), nonce, and as a means
+for applying backpressure to clients. It is chosen at random by the server for
+each response it sends.
+
+A `request` can be made without a `ticket` header if either:
+
+* It is the first `request` on a newly established connection **and** no
+  `signature` is required for the RPC
+* The client wishes to allocate additional tickets, in which case it sends a
+  `request` with all optional fields set to null
+
+Otherwise, a ticket received by any of the `response` variants (`next` field)
+must be used for subsequent RPCs. The issuer may refuse to issue more tickets by
+sending an empty response.
+
+The issuer must keep track of the issued tickets until either a `request`
+bearing the ticket is received or the connection is terminated. If a `request`
+without a `ticket` header, or with a `ticket` value which is unknown to the
+issuer, is received, the service MUST terminate the connection, optionally after
+sending an `error` response.
+
+The initiator of an RPC determines, based on application semantics, if the
+request requires authentication. If it does, the initiator requests a signature
+from the `ssh-agent` over the concatenation of
+
+    ticket || payload
+
+and includes the response, encoded as per [RFC8032], in the `signature` field.
+
+The receiver of an RPC with a non-null `signature` header MUST verify the
+`signature` against its own public key. If signature verification fails, it
+responds with a `forbidden` error, regardless of whether the RPC actually
+requires authentication as per the application semantics.
+
+
+## Architecture
+
+With the prerequisites in place, we lay out the following architecture:
```

Comment:
The `PubSub` component isn't connected to the `peer-to-peer` component. Does that mean it's not going to be used for events as part of the p2p stack?

Reply:
Right... ASCII diagrams on lenient 80 columns are hard